---
title: "CombinedSpecies"
author: "Yasaman Rezvani"
date: "10/28/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r libs}
library(openxlsx)
library(readxl)
library(tidyverse)
library(Seurat)
library(hdf5r)
library(sctransform)

```

```{r help functions}

processCounts <- function(expr){
  cols <- expr$index
  expr <- t(expr[,-1])
  colnames(expr) <- cols
  return(expr)
}

processDFtoMatrix <- function(df){
  matx <- as.matrix(df[,-1])
  rownames(matx) <- df[,1]
  return(matx)
  
}

prep_S.O <- function(S.O){
  S.O <- NormalizeData(S.O, normalization.method = "LogNormalize", scale.factor = 10000)
  S.O <- FindVariableFeatures(S.O, selection.method = "vst", nfeatures = 2000)
  all.genes <- rownames(S.O)
  S.O <- ScaleData(S.O, features = all.genes)
  S.O <- RunPCA(S.O, features = VariableFeatures(object = S.O))
  S.O <- FindNeighbors(S.O, dims = 1:13)
  S.O <- FindClusters(S.O, resolution = 0.1)
  S.O <- RunUMAP(S.O, dims = 1:13)
  return(S.O)
}
```

## Malaria blood stage SC data 
## Toxo intracellular RH384 SC data

```{r Malaria}

plasmo.path <- "~/work/SCParasites/scRNAseqPlasmodAtlas/"
toxo.path   <- "~/work/SCParasites/scRNAseqToxoAtlas/"

plasmo.expr.file <- "SS2_counts_blood_stage.csv"
toxo.expr.file   <- "rh384_raw_counts_filtered.csv"

plasmo.meta.file <- "SS2_pheno_blood_stage.csv"
toxo.meta.file   <- "rh384_meta_data.csv"
toxo.obs.file    <- "rh384_obs_filtered.csv"
```

## Reading toxo & plasmo data

```{r}

plasmo.expr <- read.csv(paste(plasmo.path, plasmo.expr.file, sep = ""))
plasmo.meta <- read.csv(paste(plasmo.path, plasmo.meta.file, sep = ""))


toxo.meta <- read.csv(paste(toxo.path, toxo.meta.file,sep = ""))
toxo.obs  <- read.csv(paste(toxo.path, toxo.obs.file, sep = ""))
toxo.obs <- toxo.obs %>% select(index, cell_cycle)
toxo.expr <- read.csv(paste(toxo.path, toxo.expr.file, sep = ""))

toxo.meta <- left_join(toxo.meta, toxo.obs, by = c("X"= "index")) 
toxo.meta[is.na(toxo.meta)] = "unknown"
```

## Orthologs

```{r}
orthologs.ID <- read.xlsx("~/work/SCParasites/GT1blastorthlgs/rec_GT1.vs.P.bergei.xlsx") 
orthologs.ID <- orthologs.ID %>% select(c(query_id,subject_id)) 
colnames(orthologs.ID) <- c("TGGT1ID", "PBERGID")

```

## 

```{r}
plasmo.expr <- left_join(orthologs.ID, plasmo.expr,by = c('PBERGID' = 'X'))
plasmo.expr <- plasmo.expr %>% select(-PBERGID) 
dim(plasmo.expr)
```
## merge  toxo & plasmo data 
## merge toxo & plasmo meta data

```{r}

colnames(toxo.expr)[1]  <- "GeneName"
colnames(plasmo.expr)[1] <- "GeneName"
toxo.plasmo.df <- left_join(toxo.expr,plasmo.expr, by = "GeneName")
toxo.plasmo.df[is.na(toxo.plasmo.df)] <- 0

toxo <- toxo.meta %>% select(c("X", "cell_cycle"))
colnames(toxo)[2] <- "CellCycle_LifeStage"
plasmo <- plasmo.meta %>% select(c("X", "ShortenedLifeStage3"))
colnames(plasmo)[2] <- "CellCycle_LifeStage"
toxo.plasmo.meta <- rbind(toxo, plasmo)
```


```{r}
toxo.plasmo <- processDFtoMatrix(toxo.plasmo.df)

```

```{r}
S.O.toxo.plasmo <- CreateSeuratObject(counts = toxo.plasmo, min.cells = 3, min.features = 200)
```

```{r}
# add life stages as a column to meta data 
meta <- S.O.toxo.plasmo@meta.data
CC_LFS <- toxo.plasmo.meta$CellCycle_LifeStage[which(toxo.plasmo.meta$X %in% colnames(S.O.toxo.plasmo))]
meta["CellCycle_LifeStage"] <- CC_LFS
meta.new <- subset(meta, select = "CellCycle_LifeStage")
S.O.toxo.plasmo <- AddMetaData(S.O.toxo.plasmo, meta.new)

```


## Create Seurat Object 

```{r}

VlnPlot(S.O.toxo.plasmo, features = c("nFeature_RNA", "nCount_RNA")
        ,ncol = 2)
FeatureScatter(S.O.toxo.plasmo, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

```



```{r}
S.O.toxo.plasmo <- subset(S.O.toxo.plasmo, subset = nFeature_RNA > 200 & nFeature_RNA < 3500 )

```

```{r}
S.O.toxo.plasmo <- SCTransform(S.O.toxo.plasmo,verbose = FALSE)
```

## add cell cycle and life stage to meta data of seurat object 


```{r}
S.O.toxo.plasmo <- RunPCA(S.O.toxo.plasmo, verbose = FALSE )
S.O.toxo.plasmo <- RunUMAP(S.O.toxo.plasmo, dims = 1:30, verbose = FALSE)

S.O.toxo.plasmo <- FindNeighbors(S.O.toxo.plasmo, dims = 1:30, verbose = FALSE)
S.O.toxo.plasmo <- FindClusters(S.O.toxo.plasmo, verbose = FALSE)
DimPlot(S.O.toxo.plasmo, label = TRUE) + NoLegend()
```



```{r}
S.O.toxo.plasmo <- prep_S.O(S.O.toxo.plasmo)
DimPlot(S.O.toxo.plasmo, reduction = "umap", group.by = "CellCycle_LifeStage")
DimPlot(S.O.toxo.plasmo, reduction = "pca")
```


```{r}

pbmc_data <- Read10X("../SingleCellAtlas/Input/hg19/")
pbmc <- CreateSeuratObject(counts = pbmc_data)
pbmc <- PercentageFeatureSet(pbmc, pattern = "^MT-", col.name = "percent.mt")
pbmc <- SCTransform(pbmc, vars.to.regress = "percent.mt", verbose = FALSE)

```

```{r}
pbmc <- RunPCA(pbmc, verbose = FALSE)
pbmc <- RunUMAP(pbmc, dims = 1:30, verbose = FALSE)

pbmc <- FindNeighbors(pbmc, dims = 1:30, verbose = FALSE)
pbmc <- FindClusters(pbmc, verbose = FALSE)
DimPlot(pbmc, label = TRUE) + NoLegend()
```




















